scipy/spatial/ckdtree.pyi

```diff
@@ -1,9 +1,221 @@
+    TypeVar,
+    Union,
+)
+
 import numpy as np
+import numpy.typing as npt
+from scipy.sparse import coo_matrix, dok_matrix
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+# TODO: Replace `ndarray` with a 1D float64 array when possible
+_BoxType = TypeVar("_BoxType", None, np.ndarray)
+
+# Copied from `numpy.typing._scalar_like._ScalarLike`
+# TODO: Expand with 0D arrays once we have shape support
+_ArrayLike0D = Union[
+    bool,
+    int,
+    float,
+    complex,
+    str,
+    bytes,
+    np.generic,
+]
+_WeightType = Union[npt.ArrayLike, Tuple[npt.ArrayLike, npt.ArrayLike]]
 
 class cKDTreeNode:
     @property
     def data_points(self) -> np.ndarray: ...
     @property
     def indices(self) -> np.ndarray: ...
 
-class cKDTree: ...
+    # Read-only attributes that are technically not properties
+    @property
+    def level(self) -> int: ...
+    @property
+    def split_dim(self) -> int: ...
+    @property
+    def children(self) -> int: ...
+    @property
+    def start_idx(self) -> int: ...
+    @property
+    def end_idx(self) -> int: ...
+    @property
+    def split(self) -> float: ...
+    @property
+    def lesser(self) -> Optional[cKDTreeNode]: ...
+    @property
+    def greater(self) -> Optional[cKDTreeNode]: ...
+
+class cKDTree(Generic[_BoxType]):
+    @property
+    def n(self) -> int: ...
+    @property
+    def m(self) -> int: ...
+    @property
+    def leafsize(self) -> int: ...
+    @property
+    def size(self) -> int: ...
+    @property
+    def tree(self) -> cKDTreeNode: ...
+
+    # Read-only attributes that are technically not properties
+    @property
+    def data(self) -> np.ndarray: ...
+    @property
+    def maxes(self) -> np.ndarray: ...
+    @property
+    def mins(self) -> np.ndarray: ...
+    @property
+    def indices(self) -> np.ndarray: ...
+    @property
+    def boxsize(self) -> _BoxType: ...
+
+    # NOTE: In practice `__init__` is used as constructor, not `__new__`.
+    # The latter gives us more flexibility in setting the generic parameter
+    # though.
+    @overload
+    def __new__(  # type: ignore[misc]
+        cls,
+        data: npt.ArrayLike,
+        leafsize: int = ...,
+        compact_nodes: bool = ...,
+        copy_data: bool = ...,
+        balanced_tree: bool = ...,
+        boxsize: None = ...,
+    ) -> cKDTree[None]: ...
+    @overload
+    def __new__(
+        cls,
+        data: npt.ArrayLike,
+        leafsize: int = ...,
+        compact_nodes: bool = ...,
+        copy_data: bool = ...,
+        balanced_tree: bool = ...,
+        boxsize: npt.ArrayLike = ...,
+    ) -> cKDTree[np.ndarray]: ...
+
+    # TODO: returns a 2-tuple of scalars if `x.ndim == 1` and `k == 1`,
+    # returns a 2-tuple of arrays otherwise
+    def query(
+        self,
+        x: npt.ArrayLike,
+        k: npt.ArrayLike = ...,
+        eps: float = ...,
+        p: float = ...,
+        distance_upper_bound: float = ...,
```

Comment:
These are user facing and then maybe we should use `DecimalNumber` introduced in #13833? I don't like these personally, but since we did it there...

Reply:
Sorry I don't understand. In which case shall I use `DecimalNumber` over `float`? Once this and #13833 is in, all of this would really deserve some quick wiki (if we don't want to put this yet in the doc).

Reply:
> In which case shall I use `DecimalNumber` over `float`?

For sure when an argument is passed to a `np.array`, but in all other situations? Honestly, I'm not sure. This is very much a question of correctness vs simplicity and I'm sure what scipy's policy is here (if there is one in the first place).
 
All in all, `DecimalNumber` _is_ more correct, but it also adds a lot of verbosity with very little additional gain.

Reply:
> For sure when an argument is passed to a `np.array`, but in all other situations? Honestly, I'm not sure. This is very much a question of correctness vs simplicity and I'm sure what scipy's policy is here (if there is one in the first place).

Well, my understanding, from our interactions here and in previous issues, was that you where the reference person here 😅 This is definitely something I will raise during our meetup.

My POV is to go for simplicity or have very clear guidelines so that it does not become untractable during review/integration. If the long term objective is to have types everywhere, then it should be possible to add type without too much overhead.

Reply:
> Well, my understanding, from our interactions here and in previous issues, was that you where the reference person here 😅

Ah, to clarify: for numpy this is pretty much the case, but for scipy on the other hand I'm not really all that involved (among other reasons because there is still _a lot_ of work to be done upstream). 
Out of curiosity, when are the weekly (?) scipy meeting usually conducted? I feel it might be useful to attend this one.

Reply:
I am very happy you asked @BvB93. I just proposed to restart the SciPy meetups and the first one in two years will happen May 17 at 16 CET. Here is some more info: https://hackmd.io/pyhudrC5TgSwdJtpPldHgQ?both
I already plan to talk about type annotations as you will see. Feel free to edit this document.
Among other things, we will decide during this meeting how often this should happen.